iothub/device/src/Transport/AmqpClientConnectionPool.cs

```diff
@@ -0,0 +1,155 @@
+
+            if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySasSingle)
+            {
+                return GetIoTHubSasSingleClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityX509)
+            {
+                return GetIoTHubX509ClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySASMux)
+            {
+                return GetIoTHubSasMuxClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityIoTHubSas)
+            {
+                return GetIoTHubSasClientConnection(deviceClientEndpointIdentity);
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+        }
+        internal void RemoveClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+
+            if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySasSingle)
+            {
+                if (singleSasPool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    singleSasPool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from singleSasPool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityX509)
+            {
+                if (x509Pool.ContainsKey(deviceClientEndpointIdentity))
+                {
+                    AmqpClientConnection amqpClientConnection;
+                    x509Pool.TryRemove(deviceClientEndpointIdentity, out amqpClientConnection);
+
+                    if (amqpClientConnection != null)
+                    {
+                        if (Logging.IsEnabled) Logging.Info(this, $"{nameof(AmqpClientConnectionPool)}.{"Removed from X509Pool:"}.{nameof(amqpClientConnection)}");
+                    }
+                }
+            }
+            if (Logging.IsEnabled) Logging.Exit(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(RemoveClientConnection)}");
+        }
+
+
+        private AmqpClientConnection GetIoTHubSasSingleClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (singleSasPool.ContainsKey(deviceClientEndpointIdentity))
+            {
+                // Error handling!!!
+                return null;
```

Comment:
Maybe return existing connection instead? It could be applied all others.

Reply:
This is not relevant anymore., as the singleSasPool has been removed completely as I am implementing the Pooling. 

---
In reply to: [251628562](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251628562) [](ancestors = 251628562)